### ISSUE TITLE
Multi Request Mutations 

### ISSUE THREAD

**Issue by datamel** [cylc/cylc-uiserver#289]
On master of `cylc-ui`, `cylc-uiserver` and `cylc-flow`, running a multi mutation request on GraphiQL mutations are coming through the incorrect number of times.

**Steps to reproduce the bug**
1. Start a workflow (mine was called `test_auth`)
2. On the GraphiQL interface in the UI, 
```
 mutation {
    pause (workflows: ["test_auth"]) {
    result
  }
 stop (workflows: ["test_auth"]) {
    result
  }

}
```
This results in a workflow log of:

![image](https://user-images.githubusercontent.com/37735232/143011684-d3f9e608-538d-4ef4-91d7-d6ff7000ddde.png)

Note that this is not just a logging error, since `Workflow is already paused` appears in the log and the queue has each mutation twice (see code diff below to reproduce this logging)

**Expected behavior**
Mutations only come through once each, as requested by the user.

**Additional context**
- I added debugging to inspect the queue (to confirm my suspicions): my diff for cylc-flow is as follows:
```
diff --git a/cylc/flow/scheduler.py b/cylc/flow/scheduler.py
index af62bada8..c8271c8a0 100644
--- a/cylc/flow/scheduler.py
+++ b/cylc/flow/scheduler.py
@@ -825,6 +825,9 @@ class Scheduler:
         if qsize <= 0:
             return
         LOG.info(f"Processing {qsize} queued command(s)")
+        import queue
+        for n in list(self.command_queue.queue):
+            LOG.warning(f"Queued item: {n}")
         while True:
             try:
                 name, args, kwargs = self.command_queue.get(False)

```
- Note that this may be a problem solved in `cylc-flow` so this issue may need transferring if the root cause is there.

- I also ran a 3 mutation request, using :
```
mutation {
    pause (workflows: ["test_auth"]) {
    result
  }
 resume (workflows: ["test_auth"]) {
    result
  }
 stop (workflows: ["test_auth"]) {
    result
  }

}

```
these came through as follows:  

![image](https://user-images.githubusercontent.com/37735232/143010709-680e369a-9d31-4e4d-8a9c-9560c6f48b23.png)
 
Tagged against 1.x for now, I expect most users will use the gui buttons rather than this interface. 

**Pull requests welcome!**
This is an Open Source project - please consider contributing a bug fix
yourself (please read `CONTRIBUTING.md` before starting any work though).

**Comment by MetRonnie** [comment]
I think this is due to the way the mutation resolvers are set up. When a (single) mutation is triggered:
1. Graphene on the cylc-uiserver end processes the mutation string
1. The resolver on the cylc-uiserver end runs and sends the **whole** mutation string to the cylc-flow end
1. Graphene on the cylc-flow end processes the mutation string
1. The resolver on the cylc-flow end runs

So for your first example I think the cylc-uiserver `pause` and `stop` resolvers each send the whole mutation string to cylc-flow, and so Graphene processes and runs the `pause` and `stop` resolvers once for each mutation string received.

Note: Anyone tackling this should wait for #271 to be fixed first as the changes will probably conflict

**Comment by dwsutherland** [comment]
I'll have a look at this one (if that's ok).

**Comment by dwsutherland** [comment]
Figured out the issue; the UIS isn't designed to have multiple mutations in one request document:
- The request is resolved by the UIS
- Each root entry (`pause`, `stop`, ... ) is resolved using the resolvers at the UIS.
- The UIS mutation resolvers pass on the graphql request to the Scheduler(s)..
- The request is passed on twice for each root entry.

I'll need to think of some elegant solution...

**Comment by oliver-sanders** [comment]
Multi requests are a GraphQL feature so this is a bug, however, one our users are unlikely to hit. The GUI and CLI both only issue one issue at a time, hence this issue is tagged against 1.x (the first post-8.0.0 UIS release), however, it would be a good one to crack.

There are some possible use cases for this. If the mutations are correctly carried out one by one in the order defined in the request document they could be used for writing macros providing compound operations. I am interested in the potential to use this to codeify operational procedures.

**Comment by dwsutherland** [comment]
> Multi requests are a GraphQL feature so this is a bug

- It's a more important feature of queries.
- It's not a feature of subscriptions.
- Mutations; probably a feature (haven't checked)... is it important?

> There are some possible use cases for this.

Then we have to construct new request documents for each mutation/root-entry, as we can't just pass the whole document on (because it undermines the authorisation).

**Comment by MetRonnie** [comment]
(Update to my earlier comment: no need to wait for #271, that one is lower priority)

**Comment by oliver-sanders** [comment]
So, I think we need to parse the request document, find the mutation that matches the one being actioned (easier said than done) then drop the remainder from the document and submit it.

**Comment by oliver-sanders** [comment]
Simplest solution ATM might be to check the request document prior to forwarding the mutation to the scheduler.

If there is more than one mutation declaration in the document, raise an exception *before* forwarding the request to the scheduler. This would allow us to kick this issue down the line a bit?
